composer/trainer/trainer.py

```diff
@@ -47,6 +47,35 @@
 
 log = logging.getLogger(__name__)
 
+def _dynamic_microbatch_wrapper(func: Callable) -> Callable:
+    """Wraps function to catch CUDA Out of Memory Errors and adaptively change microbatch
+    size if enabled to miaximize GPU usage.
+    """
+    def wrapper(*args, **kwargs):
+        self = args[0]
+        # Skip adaption if not enabled
+        if not self.adaptive_grad_accum:
+            return func
+        rerun = True
+        while rerun:
+            try:
+                rerun = False
+                return func(*args, **kwargs)
+            except RuntimeError as e:
+                if "CUDA out of memory" in str(e):
```

Comment:
Maybe put this in a function called `def _is_cuda_oom` just to minimize hardcoding in this function